src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+}
+
+void
+sol_iio_close(struct sol_iio_device *device)
+{
+    int i;
+    struct sol_iio_channel *channel;
+
+    SOL_PTR_VECTOR_FOREACH_REVERSE_IDX (&device->channels, channel, i) {
+        iio_del_channel(channel);
+        sol_ptr_vector_del(&device->channels, i);
+    }
+
+    if (device->fd_handler) sol_fd_del(device->fd_handler);
+    if (device->fd > -1) close(device->fd);
+    free(device->trigger_name);
+    free(device);
+}
+
+static bool
+read_channel_index_in_buffer(struct sol_iio_channel *channel)
+{
+    struct sol_iio_device *device = channel->device;
+    char path[PATH_MAX + 1];
+    int len;
+
+    len = snprintf(path, sizeof(path), CHANNEL_SCAN_INDEX_PATH,
+        device->device_id, channel->name);
+    if (len < 0 || sol_util_read_file(path, "%d", &channel->index) < 0)
+        return false;
+
+    return true;
+}
+
+struct sol_iio_channel *
+sol_iio_add_channel(struct sol_iio_device *device, const char *name)
+{
+    char path[PATH_MAX + 1];
+    struct sol_iio_channel *channel;
+    struct stat st;
+    int len;
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        name);
+    if (len < 0 || len > PATH_MAX || stat(path, &st)) {
+        SOL_WRN("Could not find channel [%s] for device%d", name,
+            device->device_id);
+        return NULL;
+    }
+
+    channel = calloc(1, sizeof(struct sol_iio_channel));
+    SOL_NULL_CHECK(channel, NULL);
+
+    channel->name = strdup(name);
+    channel->device = device;
+
+    channel_get_scale(channel);
+    channel_get_offset(channel);
+
+    if (device->buffer_enabled) {
+        if (!enable_channel_scan(channel)) {
+            SOL_WRN("Could not enable scanning of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_scan_info(channel)) {
+            SOL_WRN("Could not read scanning info of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_index_in_buffer(channel)) {
+            SOL_WRN("Could not read index in buffer of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        channel->offset_in_buffer = -1;
+    }
+
+    sol_ptr_vector_append(&channel->device->channels, channel);
+
+    SOL_DBG("channel %s added. scale: %lf - offset: %d - storagebits: %d",
+        channel->name, channel->scale, channel->offset, channel->storagebits);
+
+    return channel;
+
+error:
+    iio_del_channel(channel);
+
+    return NULL;
+}
+
+bool
+sol_iio_read_channel_value(struct sol_iio_channel *channel, double *value)
+{
+    int len, raw_value;
+    char path[PATH_MAX + 1];
+    struct sol_iio_device *device = channel->device;
+
+    if (device->buffer_enabled) {
+        SOL_WRN("device%d is using buffer - use iio_read_buffer_channel_value() instead",
+            device->device_id);
```

Comment:
As I said in other parts you should be keeping the buffer anyways, for performance reason. And here you can help the user, so he doesn't need to account himself 2 code paths...